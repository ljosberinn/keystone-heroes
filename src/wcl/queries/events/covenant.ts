import type { CastEvent, AllTrackedEventTypes } from "./types";
import { createIsSpecificEvent } from "./utils";

export const SHARED_COVENANT_ABILITIES = {
  FLESHCRAFT: 324_631,
  SOULSHAPE: 310_143,
  SUMMON_STEWARD: 324_739,
  DOOR_OF_SHADOWS: 300_728,
} as const;

export const generalCovenantExpression = `type in ("cast", "applydebuff") and ability.id in (${Object.values(
  SHARED_COVENANT_ABILITIES
).join(", ")})`;

const isSharedCovenantAbility = createIsSpecificEvent<CastEvent>({
  abilityGameID: Object.values(SHARED_COVENANT_ABILITIES),
  type: "cast",
});

export const filterCovenantCastEvents = (
  allEvents: AllTrackedEventTypes[]
): CastEvent[] => {
  return allEvents.filter(isSharedCovenantAbility);
};
