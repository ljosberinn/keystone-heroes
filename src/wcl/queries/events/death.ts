import { EXPLOSIVE_HEALTH } from "./affixes/explosive";
import type {
  AllTrackedEventTypes,
  DeathEvent,
  BeginCastEvent,
  DamageEvent,
} from "./types";

const RENEWING_MIST = 300_155;
/**
 * @description filters for
 * - _any_ death event
 * - the `Renewing Mist` cast Tirnenn Villagers do upon their "death"
 */
export const deathFilterExpression = `(type = "death" and target.type != "pet") or (ability.id = ${RENEWING_MIST} and type = "begincast")`;
// TODO: feign false doesnt work?
// export const friendliesDeathFilterExpression =
//   'target.type = "player" and type = "death"'; //  and feign = false

export const filterEnemyDeathEvents = (
  allEvents: AllTrackedEventTypes[],
  actorIDSet: Set<number>,
  explosiveTargetID: number | null
): (DeathEvent | BeginCastEvent | DamageEvent)[] => {
  return allEvents.filter(
    (event): event is DeathEvent | BeginCastEvent | DamageEvent => {
      if (event.type === "death") {
        // may not be an actor, aka player
        // source of death must be environment
        return !actorIDSet.has(event.targetID) && event.sourceID === -1;
      }

      // see Renewing Mist
      if (event.type === "begincast") {
        return event.abilityGameID === RENEWING_MIST;
      }

      if (explosiveTargetID && event.type === "damage") {
        // explosives do not send death events
        // however, `overkill` is set given non melee hits
        // on melee hits, there's no `overkill` but the amount must be higher than 224
        return (
          event.targetID === explosiveTargetID &&
          (event.overkill !== undefined || event.amount >= EXPLOSIVE_HEALTH)
        );
      }

      return false;
    }
  );
};

export const filterPlayerDeathEvents = (
  allEvents: AllTrackedEventTypes[],
  actorIDSet: Set<number>
  // remarkableSpellEvents: CastEvent[]
): DeathEvent[] => {
  // const actorIDSet = new Set(
  // playerMetaInformation.map((dataset) => dataset.actorID)
  // );

  // const hunter = playerMetaInformation.find(
  //   (player) => player.class === PlayableClass.Hunter
  // );

  return allEvents.filter((event): event is DeathEvent => {
    return (
      event.type === "death" &&
      // must be a known actor
      actorIDSet.has(event.targetID) &&
      // source of death must be environment
      event.sourceID === -1
    );
  });

  // if (!hunter) {
  //   return deathEvents;
  // }

  // const hunterDeaths = deathEvents.filter(
  //   (event) => event.targetID === hunter.actorID
  // );

  // if (hunterDeaths.length === 0) {
  //   return deathEvents;
  // }

  // return deathEvents.filter((event) => {
  //   const isHunterDeath = hunterDeaths.includes(event);

  //   if (!isHunterDeath) {
  //     return true;
  //   }

  //   const nextHunterCD = remarkableSpellEvents.find((e) => {
  //     return e.sourceID === event.targetID && e.timestamp > event.timestamp;
  //   });

  //   if (!nextHunterCD) {
  //     return true;
  //   }

  //   // assume a hunter feigned if he used a cd within the next 2 seconds
  //   // could alternatively be solved by querying
  //   // hostilityType: Friendlies, dataType: Deaths
  //   // once separately...
  //   return nextHunterCD.timestamp - event.timestamp >= 2000;
  // });
};
