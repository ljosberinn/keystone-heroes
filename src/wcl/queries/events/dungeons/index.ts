import { DungeonIDs } from "../../../../db/data/dungeons";
import type { AllTrackedEventTypes } from "../../events/types";
import { filterExpression as dosFilterExpression, getDOSEvents } from "./dos";
import { filterExpression as hoaFilterExpression, getHOAEvents } from "./hoa";
import {
  filterExpression as motsFilterExpression,
  getMOTSEvents,
} from "./mots";
import { filterExpression as nwFilterExpression, getNWEvents } from "./nw";
import { filterExpression as pfFilterExpression, getPFEvents } from "./pf";
import { filterExpression as sdFilterExpression, getSDEvents } from "./sd";
import { filterExpression as soaFilterExpression, getSOAEvents } from "./soa";
import { filterExpression as topFilterExpression, getTOPEvents } from "./top";

const dungeonExpressionMap: Record<DungeonIDs, string[]> = {
  [DungeonIDs.DE_OTHER_SIDE]: dosFilterExpression,
  [DungeonIDs.HALLS_OF_ATONEMENT]: hoaFilterExpression,
  [DungeonIDs.PLAGUEFALL]: pfFilterExpression,
  [DungeonIDs.SANGUINE_DEPTHS]: sdFilterExpression,
  [DungeonIDs.SPIRES_OF_ASCENSION]: soaFilterExpression,
  [DungeonIDs.THEATER_OF_PAIN]: topFilterExpression,
  [DungeonIDs.THE_NECROTIC_WAKE]: nwFilterExpression,
  [DungeonIDs.MISTS_OF_TIRNA_SCITHE]: motsFilterExpression,
};

export const getDungeonExpression = (id: DungeonIDs): string[] =>
  dungeonExpressionMap[id];

export const filterDungeonEvents = (
  allEvents: AllTrackedEventTypes[],
  dungeonID: DungeonIDs,
  actorIDSet: Set<number>
): ReturnType<
  | typeof getDOSEvents
  | typeof getHOAEvents
  | typeof getPFEvents
  | typeof getSDEvents
  | typeof getSOAEvents
  | typeof getTOPEvents
  | typeof getNWEvents
  | typeof getMOTSEvents
> => {
  switch (dungeonID) {
    case DungeonIDs.DE_OTHER_SIDE:
      return getDOSEvents(allEvents);
    case DungeonIDs.HALLS_OF_ATONEMENT:
      return getHOAEvents(allEvents);
    case DungeonIDs.PLAGUEFALL:
      return getPFEvents(allEvents, actorIDSet);
    case DungeonIDs.SANGUINE_DEPTHS:
      return getSDEvents(allEvents);
    case DungeonIDs.SPIRES_OF_ASCENSION:
      return getSOAEvents(allEvents);
    case DungeonIDs.THEATER_OF_PAIN:
      return getTOPEvents(allEvents);
    case DungeonIDs.THE_NECROTIC_WAKE:
      return getNWEvents(allEvents);
    case DungeonIDs.MISTS_OF_TIRNA_SCITHE:
      return getMOTSEvents(allEvents);
    default:
      return [];
  }
};
