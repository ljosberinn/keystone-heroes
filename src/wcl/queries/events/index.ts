import { Affixes } from "@prisma/client";

import type { DungeonIDs } from "../../../db/data/dungeons";
import type {
  AllTrackedEventTypes,
  BeginCastEvent,
  DamageEvent,
  DeathEvent,
} from "../events/types";
import { filterAffixEvents, getAffixExpression } from "./affixes";
import { findExplosiveTargetID } from "./affixes/explosive";
import {
  cheatDeathFilterExpression,
  filterCheatDeathEvents,
} from "./cheathDeath";
import {
  generalCovenantExpression,
  filterCovenantCastEvents,
} from "./covenant";
import {
  deathFilterExpression,
  filterEnemyDeathEvents,
  filterPlayerDeathEvents,
} from "./death";
import { getDungeonExpression, filterDungeonEvents } from "./dungeons";
import {
  filterPlayerInterruptEvents,
  interruptFilterExpression,
} from "./interrupts";
import {
  engineeringBattleRezExpression,
  invisibilityFilterExpression,
  filterProfessionEvents,
} from "./professions";
import {
  filterRemarkableSpellEvents,
  remarkableSpellFilterExpression,
} from "./spells";
import { filterTrinkets, trinketsFilterExpression } from "./trinkets";
import { recursiveGetEvents } from "./utils";

export type EventParams = {
  reportID: string;
  startTime: number;
  endTime: number;
  dungeonID: DungeonIDs;
  fightID: number;
  affixes: Affixes[];
};

const generateFilterExpression = ({
  dungeonID,
  affixes,
}: Pick<EventParams, "dungeonID" | "affixes">) => {
  return [
    deathFilterExpression,
    remarkableSpellFilterExpression,
    engineeringBattleRezExpression,
    invisibilityFilterExpression,
    interruptFilterExpression,
    generalCovenantExpression,
    cheatDeathFilterExpression,
    trinketsFilterExpression,
    ...getDungeonExpression(dungeonID),
    ...getAffixExpression(affixes),
  ]
    .flat()
    .map((part) => `(${part})`)
    .join(" or ");
};

export const getEvents = async (
  params: EventParams,
  playerMetaInformation: { actorID: number; class: number }[]
): Promise<{
  allEvents: AllTrackedEventTypes[];
  playerDeathEvents: DeathEvent[];
  enemyDeathEvents: (DeathEvent | BeginCastEvent | DamageEvent)[];
  explosiveTargetID: number | null;
}> => {
  const filterExpression = generateFilterExpression({
    dungeonID: params.dungeonID,
    affixes: params.affixes,
  });

  const allEvents = await recursiveGetEvents<AllTrackedEventTypes>({
    startTime: params.startTime,
    endTime: params.endTime,
    reportID: params.reportID,
    filterExpression,
  });

  const actorIDSet = new Set(
    playerMetaInformation.map((dataset) => dataset.actorID)
  );

  const dungeonEvents = filterDungeonEvents(
    allEvents,
    params.dungeonID,
    actorIDSet
  );
  const { explosiveTargetID, events: affixEvents } = filterAffixEvents(
    allEvents,
    params.affixes
  );
  const professionEvents = filterProfessionEvents(allEvents);
  const remarkableSpellEvents = filterRemarkableSpellEvents(allEvents);
  const playerDeathEvents = filterPlayerDeathEvents(
    allEvents,
    actorIDSet
    // remarkableSpellEvents
  );
  const interruptEvents = filterPlayerInterruptEvents(allEvents);
  const enemyDeathEvents = filterEnemyDeathEvents(
    allEvents,
    actorIDSet,
    params.affixes.includes(Affixes.Explosive)
      ? findExplosiveTargetID(allEvents)
      : null
  );
  const sharedCovenantAbilitiesEvents = filterCovenantCastEvents(allEvents);
  const cheatDeathEvents = filterCheatDeathEvents(allEvents);
  const trinketEvents = filterTrinkets(allEvents);

  return {
    allEvents: [
      ...professionEvents,
      ...dungeonEvents,
      ...affixEvents,
      ...remarkableSpellEvents,
      ...playerDeathEvents,
      ...interruptEvents,
      ...sharedCovenantAbilitiesEvents,
      ...cheatDeathEvents,
      ...trinketEvents,
    ].sort((a, b) => a.timestamp - b.timestamp),
    playerDeathEvents,
    enemyDeathEvents,
    explosiveTargetID,
  };
};
