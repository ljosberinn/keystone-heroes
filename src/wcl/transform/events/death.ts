import { EventType } from "@prisma/client";

import { PF } from "../../queries/events/dungeons/pf";
import type { DeathEvent } from "../../queries/events/types";
import type { PersistedDungeonPull, Processor } from "../utils";

export const deathProcessor: Processor<
  DeathEvent,
  { pull: PersistedDungeonPull }
> = (event, { targetNPCID, targetPlayerID, pull }) => {
  const {
    timestamp,
    killerInstance: sourceNPCInstance = null,
    killerID,
    targetInstance: targetNPCInstance = null,
  } = event;

  // player death
  if (targetPlayerID) {
    const sourceNPCID = killerID
      ? pull.enemyNPCs.find((npc) => npc.id === killerID)?.gameID ?? null
      : null;

    return {
      timestamp,
      eventType: EventType.Death,
      sourceNPCInstance,
      targetPlayerID,
      sourceNPCID,
    };
  }

  if (
    PF.GREEN_BUFF.unit === targetNPCID ||
    PF.RED_BUFF.unit === targetNPCID ||
    PF.PURPLE_BUFF.unit === targetNPCID
  ) {
    return {
      timestamp,
      eventType: EventType.Death,
      targetNPCInstance,
      targetNPCID,
    };
  }

  return null;
};
