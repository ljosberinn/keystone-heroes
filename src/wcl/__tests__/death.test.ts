import { PF } from "../queries/events/dungeons/pf";
import type { DeathEvent } from "../queries/events/types";
import { deathProcessor } from "../transform/events/death";
import allEvents from "./fixtures/allEvents.json";

describe("death", () => {
  const params = {
    targetPlayerID: null,
    targetNPCID: null,
    sourcePlayerID: null,
    sourceNPCID: null,
    pull: {
      enemyNPCs: [],
      startTime: 0,
      endTime: 0,
      id: 1,
      x: 0,
      y: 0,
      isWipe: false,
      maps: [],
      percent: 0,
    },
  };

  const deathEvents = allEvents.filter(
    (event): event is DeathEvent => event.type === "death"
  );

  test("tracks player death events", () => {
    const result = deathEvents.map((event) =>
      deathProcessor(event, { ...params, targetPlayerID: 1 })
    );

    expect(result).toMatchSnapshot();
  });

  test("tracks PF Slime Death events", () => {
    const result = [
      PF.GREEN_BUFF.unit,
      PF.PURPLE_BUFF.unit,
      PF.RED_BUFF.unit,
    ].flatMap((targetNPCID) =>
      deathEvents.map((event) =>
        deathProcessor(event, { ...params, targetNPCID })
      )
    );

    expect(result).toMatchSnapshot();
  });

  test("returns null for unknown death events", () => {
    expect(
      deathProcessor(
        {
          type: "death",
          abilityGameID: 1,
          sourceID: 1,
          targetID: 0,
          timestamp: 1,
        },
        params
      )
    ).toBeNull();
  });
});
