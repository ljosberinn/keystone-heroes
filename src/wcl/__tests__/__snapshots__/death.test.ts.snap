// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`death tracks PF Slime Death events 1`] = `
[
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3186763,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3331087,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3424154,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3436647,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3489087,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3506311,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3512625,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3514282,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3517463,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3534137,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3559079,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3560518,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3563937,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3566890,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3568140,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3573319,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3576106,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3793055,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3889770,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 3910966,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 4036935,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 4273729,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": null,
    "timestamp": 4634252,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": 4,
    "timestamp": 9399646,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": 5,
    "timestamp": 9400225,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": 1,
    "timestamp": 9400225,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": 6,
    "timestamp": 9430006,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": 1,
    "timestamp": 9486481,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": 2,
    "timestamp": 9769975,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": 2,
    "timestamp": 9770061,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": 3,
    "timestamp": 11090607,
  },
  {
    "eventType": "Death",
    "targetNPCID": 163891,
    "targetNPCInstance": 7,
    "timestamp": 11099140,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3186763,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3331087,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3424154,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3436647,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3489087,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3506311,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3512625,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3514282,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3517463,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3534137,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3559079,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3560518,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3563937,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3566890,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3568140,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3573319,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3576106,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3793055,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3889770,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 3910966,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 4036935,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 4273729,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": null,
    "timestamp": 4634252,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": 4,
    "timestamp": 9399646,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": 5,
    "timestamp": 9400225,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": 1,
    "timestamp": 9400225,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": 6,
    "timestamp": 9430006,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": 1,
    "timestamp": 9486481,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": 2,
    "timestamp": 9769975,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": 2,
    "timestamp": 9770061,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": 3,
    "timestamp": 11090607,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164707,
    "targetNPCInstance": 7,
    "timestamp": 11099140,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3186763,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3331087,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3424154,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3436647,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3489087,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3506311,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3512625,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3514282,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3517463,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3534137,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3559079,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3560518,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3563937,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3566890,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3568140,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3573319,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3576106,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3793055,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3889770,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 3910966,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 4036935,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 4273729,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": null,
    "timestamp": 4634252,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": 4,
    "timestamp": 9399646,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": 5,
    "timestamp": 9400225,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": 1,
    "timestamp": 9400225,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": 6,
    "timestamp": 9430006,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": 1,
    "timestamp": 9486481,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": 2,
    "timestamp": 9769975,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": 2,
    "timestamp": 9770061,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": 3,
    "timestamp": 11090607,
  },
  {
    "eventType": "Death",
    "targetNPCID": 164705,
    "targetNPCInstance": 7,
    "timestamp": 11099140,
  },
]
`;

exports[`death tracks player death events 1`] = `
[
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3186763,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3331087,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3424154,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3436647,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3489087,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3506311,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3512625,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3514282,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3517463,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3534137,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3559079,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3560518,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3563937,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3566890,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3568140,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3573319,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3576106,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3793055,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3889770,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 3910966,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 4036935,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 4273729,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 4634252,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 9399646,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 9400225,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 9400225,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 9430006,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 9486481,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 9769975,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 9770061,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 11090607,
  },
  {
    "eventType": "Death",
    "sourceNPCID": null,
    "sourceNPCInstance": null,
    "targetPlayerID": 1,
    "timestamp": 11099140,
  },
]
`;
